### DISCUSSION TITLE
[expo-notifications] Notifications.requestPermissionsAsync doesn't show permissions prompt

### DISCUSSION THREAD

**Discussion by alexanderdavide** [expo/expo#17550] [Archived: Questions]
Hi,

using the code from [Push Notifications Overview](https://docs.expo.dev/push-notifications/overview/), I assume `Notifications.requestPermissionsAsync()` to show a prompt requesting permissions as described by the [docs](https://docs.expo.dev/versions/latest/sdk/notifications/#requestpermissionsasyncrequest-notificationpermissionsrequest-promisenotificationpermissionsstatus). Against my expectations, no prompt is shown.

What also is strange to me, the [docs](https://docs.expo.dev/push-notifications/overview/#testing) claim the following which is not true in my experiments:
> Additionally, when calling Notifications.requestPermissionsAsync() on the simulator, it will resolve immediately with undetermined as the status, regardless of whether you choose to allow or not.

The permission status is granted by default, without any prompt like explained. The only exception is API level 33 which denies the permissions by default, complying with changes since [Android 13](https://developer.android.com/about/versions/13/changes/notification-permission) but it neither shows a prompt.

I'm on bare workflow but still running on Expo Go. I've tried uninstalling Expo Go multiple times so there also is no initial prompt which remembers the decision. I've also ran application builds directly on the devices but it didn't make a difference. Following the devices I've tested with the respective results:

| API level | Device Type | Shows permission prompt | Permission status |
|-----------|-------------|-------------------------|-------------------|
| 29        | physical    | no                      | granted           |
| 31        | emulator    | no                      | granted           |
| 32        | emulator    | no                      | granted           |
| 33        | emulator    | no                      | denied            |

I'm wondering if I misunderstand the concept how permissions are working here or if there's a problem with expo-notifications.  I'd greatly appreciate you getting me on the right track.

**Comment by harryfino** [comment]
@alexanderdavide I'm experiencing the same thing. Have you learned anything on fixed this issue since May?

**Reply by alexanderdavide** [reply]
I haven't. I find myself in a early stage of development which is why I haven't bothered too much since then. My current targetSdkVersion is 31 and notifications are working on the devices I've tested. However, I assume that this topic will become a problem further down the line.

**Comment by viljark** [comment]
I had the same issue, I could not trigger a notification permission prompt at all with `await Notifications.requestPermissionsAsync();` on Android 13.

So the workaround for me was registering the notification channel with `Notifications.setNotificationChannelAsync` when I needed the prompt to be displayed. I don't do any permission checking before registering the channel (as it is in the [example](https://docs.expo.dev/versions/latest/sdk/notifications/#api))

**Reply by harryfino** [reply]
Thank you so much for posting this. I've wasted about 72 hours researching this and came up with nothing until I tried your workaround. How can I buy you a coffee?

**Reply by viljark** [reply]
> Thank you so much for posting this. I've wasted about 72 hours researching this and came up with nothing until I tried your workaround. How can I buy you a coffee?

Glad I was able to help, you can buy me a coffee [here](https://ko-fi.com/viljark) 😄☕️

**Reply by ndjerrou** [reply]
Can we see your code ? Did u what u explained but still no pop up

**Reply by viljark** [reply]
> Notifications.setNotificationChannelAsync

I have this in my App.ts, right after imports.

```
if (Platform.OS === 'android') {
	Notifications.setNotificationChannelAsync('default', {
		name: 'default',
		importance: Notifications.AndroidImportance.MAX,
		vibrationPattern: [0, 250, 250, 250],
		enableVibrate: true,
		enableLights: true,
	})
}
```

**Reply by ndjerrou** [reply]
Check this code : 
```

import AsyncStorage from '@react-native-async-storage/async-storage';
import * as Notifications from 'expo-notifications';
import { useEffect } from 'react';
import { StyleSheet, Text, View } from 'react-native';

Notifications.setNotificationChannelAsync('default', {
  name: 'default',
  importance: Notifications.AndroidImportance.MAX,
  vibrationPattern: [0, 250, 250, 250],
  lightColor: '#FF231F7C',
});

export default function App() {
  const askPermission = async () => {
    const { status } = await Notifications.requestPermissionsAsync();
    console.log(status);
  };

  useEffect(() => {
    askPermission();
  }, []);


  AsyncStorage.getAllKeys().then((keys) => console.log('', keys));

  return (
    <View style={styles.container}>
      <Text>Testing Permissions...</Text>
    </View>
  );
}

```

No prompt is coming, using my phone as an emulator on Android :(

**Reply by viljark** [reply]
Push notifications don't work in simulators. You have to use real device.

https://docs.expo.dev/versions/latest/sdk/notifications/#api

**Reply by ndjerrou** [reply]
My phone is considerated as a real device no?

Using expo go

**Reply by viljark** [reply]
Oh, understood you wrong. Verify you don't have notifications disabled in app settings. You could also build an apk to check if has the same issue. Expo go sometimes behaves differently than built apk.

**Reply by rikur** [reply]
Having the same issue on Android 13 devices. The permission popup only shows up after the app is killed and opened again.

My workaround is to not ask for permission for Android 13 devices for the time being as the notifications work without it.

**Comment by gustavoccintrao** [comment]
I'm having this issue on iOS too, no prompt is show, any help? SDK 44.

**Reply by ndjerrou** [reply]
It works like a charm in IOS. Did u use requestPermissionsAsync ?

**Reply by gustavoccintrao** [reply]
Yes, I’m using requestPermissionsAsync. I logged the “{status}” and received an insta denied. Replacing the “{status}” for a simple “result” const to check the other properties led me to find that canAskAgain is false. It seems that if you already disabled notifications once in iOS you can’t reset the status to show the prompt without uninstalling Expo and installing it again.

**Reply by viljark** [reply]
There are 2 UX workarounds I have used in similar cases.

First option is to display your own dialog, and if user agrees that they want notification, then request the permission. If they don't agree with your dialog, then you skip the permission asking and can re-trigger it when it might make sense again and user is more willing to accept the permission.

Second workaround when `canAskAgain` is already false, is to display a custom prompt again with custom message and a button, which will call `Linking.openURL('app-settings:')`, there user can allow the notificaiton manually.

**Reply by gustavoccintrao** [reply]
I ended up using the second workaround, thanks!

**Comment by valentimcanejo** [comment]
I'm facing the same issue, I tried literally everythin and the prompt asking the user to accept notifications doesn't show, I followed the whole expo-notifications documentation and didn't work.

**Comment by mamadouD29** [comment]
Hey, guys. 
did anybody find a solution or a workaround for this issue ? it is a little bit confusing to just grant the permission without user interaction.

**Comment by t0ma5h** [comment]
For me, the first time `canAskAgain` is true, and if you just back out from the modal on android, it turns to false, even though it shouldn't. Untill you press `Don't allow`, `canAskAgain` does not change to `true` even though `requestPermissionsAsync` opens the native modal perfectly fine. Once you press `Don't allow`, `canAskAgain` turns to true and then it works ok the second time.

**Comment by MOHAMED-LAAGUILI** [comment]
same here can find a solution that automaticly request permision till now i do it  manually from app
